node-repository/src/main/java/com/yahoo/vespa/hosted/provision/backup/Snapshot.java

```diff
@@ -1,36 +1,86 @@
 package com.yahoo.vespa.hosted.provision.backup;
 
+import com.google.common.collect.ImmutableMap;
 import com.yahoo.config.provision.ClusterSpec;
 import com.yahoo.config.provision.HostName;
 import com.yahoo.vespa.hosted.provision.node.ClusterId;
 
 import java.time.Instant;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.UUID;
 
 /**
  * A backup snapshot of a node's local data. Only {@link ClusterSpec.Type#content} nodes support snapshots.
  *
  * @author mpolden
  */
-public record Snapshot(SnapshotId id, HostName hostname, State state, Instant createdAt, ClusterId cluster, int clusterIndex) {
+public record Snapshot(SnapshotId id, HostName hostname, State state, History history, ClusterId cluster, int clusterIndex) {
 
     public Snapshot {
         Objects.requireNonNull(id);
         Objects.requireNonNull(state);
         Objects.requireNonNull(hostname);
-        Objects.requireNonNull(createdAt);
+        Objects.requireNonNull(history);
         Objects.requireNonNull(cluster);
         if (clusterIndex < 0) {
             throw new IllegalArgumentException("clusterIndex cannot be negative, got " + cluster);
         }
     }
 
-    public Snapshot with(State state) {
-        if (state.compareTo(this.state) < 0) {
+    public Snapshot with(State state, Instant at) {
+        if (!canChangeTo(state)) {
             throw new IllegalArgumentException("Cannot change state of " + this + " to " + state);
         }
-        return new Snapshot(id, hostname, state, createdAt, cluster, clusterIndex);
+        return new Snapshot(id, hostname, state, history.with(state, at), cluster, clusterIndex);
+    }
+
+    private boolean canChangeTo(State state) {
+        // Allow repeated restores
+        if (state == State.restoring && this.state == State.restored) return true;
```

Comment:
This doesn't sound like something we want to always allow?
Perhaps an operator should force-change the state away from restored first?

Reply:
You mean restoring->restored->created->restoring? I'll think about it. Right now I want to allow this for testing purposes.